File: benchmarks/compare/raptor/go/service.go
package raptor

// Service models a GTFS calendar with include/exclude date overrides.
//
// The dates index distinguishes "key present" from "truthy" exactly like the
// reference: a present key with value true includes that date (and
// short-circuits), a present key with value false excludes it, an absent key
// falls through to the start/end/day-of-week check.
type Service struct {
	StartDate DateNumber
	EndDate   DateNumber
	Days      map[DayOfWeek]bool
	Dates     map[DateNumber]bool
}

// NewService builds a Service.
func NewService(startDate, endDate DateNumber, days map[DayOfWeek]bool, dates map[DateNumber]bool) *Service {
	if days == nil {
		days = map[DayOfWeek]bool{}
	}
	if dates == nil {
		dates = map[DateNumber]bool{}
	}
	return &Service{StartDate: startDate, EndDate: endDate, Days: days, Dates: dates}
}

// RunsOn replicates:
//
//	dates[date] === true OR
//	  (!hasOwn(dates, date) && startDate <= date <= endDate && days[dow])
func (s *Service) RunsOn(date DateNumber, dow DayOfWeek) bool {
	if v, ok := s.Dates[date]; ok {
		// Key present: its truthiness is the value. The JS `||` short-circuits
		// here regardless of whether it is true or false (a false value makes
		// the !hasOwn clause false, so the whole expression is false).
		return v
	}

	return s.StartDate <= date && s.EndDate >= date && s.Days[dow]
}

File: benchmarks/compare/raptor/rust/src/service.rs
//! Calendar service logic (port of src/gtfs/Service.ts).

use std::collections::HashMap;

use crate::gtfs::{DateNumber, DayOfWeek};

/// GTFS calendar service.
#[derive(Debug, Clone)]
pub struct Service {
    start_date: DateNumber,
    end_date: DateNumber,
    /// days[dow] = runs on that day of week.
    days: HashMap<DayOfWeek, bool>,
    /// The include/exclude index: a key present with value `true` = include,
    /// present with `false` = exclude. The key-present-vs-truthy distinction is
    /// load-bearing (contract trap #7).
    dates: HashMap<DateNumber, bool>,
}

impl Service {
    pub fn new(
        start_date: DateNumber,
        end_date: DateNumber,
        days: HashMap<DayOfWeek, bool>,
        dates: HashMap<DateNumber, bool>,
    ) -> Self {
        Service {
            start_date,
            end_date,
            days,
            dates,
        }
    }

    /// Mirrors the JS:
    /// ```text
    /// dates[date] === true OR
    /// ( !hasOwn(dates, date) && startDate <= date && endDate >= date && days[dow] )
    /// ```
    pub fn runs_on(&self, date: DateNumber, dow: DayOfWeek) -> bool {
        // `this.dates[date]` is truthy only when the key is present with value true.
        if let Some(&true) = self.dates.get(&date) {
            return true;
        }

        !self.dates.contains_key(&date)
            && self.start_date <= date
            && self.end_date >= date
            && *self.days.get(&dow).unwrap_or(&false)
    }
}
